### PR TITLE
Cleanup updateMembershipBasedOnCompletionOfContribution() to use API4 and deprecate functions

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
First attempt at a cleanup of the `CRM_Contribute_BAO_Contribution::updateMembershipBasedOnCompletionOfContribution()` function.

1. Switch direct SQL calls to API4.
2. Replace CRM_Utils_Value::array() with ??

Before
----------------------------------------
Older code

After
----------------------------------------
Newer code

Technical Details
----------------------------------------
This should be "no-change" functionally. Per comments there is probably quite a bit of simplification/clean up that could be done here. This is a first step to refactoring and we get to deprecate another function :-)

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -4188,30 +4190,29 @@ public static function updateMembershipBasedOnCompletionOfContribution($contribu
...
-ORDER BY  id DESC
-LIMIT 1;";
-      $dao = CRM_Core_DAO::executeQuery($sql);
-      if ($dao->fetch()) {
-        if (!empty($dao->membership_type_id)) {
-          $membershipParams['membership_type_id'] = $dao->membership_type_id;
-        }
+      $preChangeMembership = MembershipLog::get(FALSE)
+        ->addSelect('membership_type_id')
+        ->addWhere('membership_id', '=', 65)
```
Comment (colemanw) [inline CRM/Contribute/BAO/Contribution.php:4195]:
```suggestion
        ->addWhere('membership_id', '=', $membershipParams['id'])
```

Comment (mattwire) [reply CRM/Contribute/BAO/Contribution.php:4195]: But isn't every membership ID 65? Oh, I see. It should be 42.

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
@eileenmcnaughton Tests passing now - I had to make a test change because step debugging showed that LineItem::get() was still retrieving the unmodified LineItems and not the ones updated by direct SQL query.

**Comment by eileenmcnaughton** [comment]
@mattwire that is super weird - cos `LineItem` is a generic api - @colemanw @totten any idea why it might be - see @mattwire's last commit

**Comment by mattwire** [comment]
> @mattwire that is super weird - cos `LineItem` is a generic api - @colemanw @totten any idea why it might be - see @mattwire's last commit
> 
> This must be the get in question ?
> 
> https://github.com/civicrm/civicrm-core/blob/4f30edd2ea8ca7395fb4ede36b06231fd10021ca/CRM/Financial/BAO/Order.php#L1298

No, it's here: https://github.com/civicrm/civicrm-core/pull/28605/files#diff-4c9d0b1abe07057a4eea2b47bc627eecb95face8ed8d86c1c005312a52cca811R4209

ie. the call to LineItem::get() that I added to replace a direct SQL call. Running through xdebug showed price_field_value=3 instead of expected 1 (but those IDs may change depending on test environment/order). So it definitely *wasn't* picking up the change made by https://github.com/civicrm/civicrm-core/pull/28605/files#diff-f854cc50e25729d771411105b2ea81397cc1946abb90e79be703a3ec5c6e5bb6R3572

**Comment by mattwire** [comment]
@eileenmcnaughton If tests pass this should be good to go now. I've identified and fixed the LineItem wierdness in the test.

File: CRM/Contribute/BAO/Contribution.php [hunk 2]
```diff
@@ -4253,7 +4256,7 @@ public static function updateMembershipBasedOnCompletionOfContribution($contribu
         );
 
         unset($dates['end_date']);
-        $membershipParams['status_id'] = CRM_Utils_Array::value('id', $calcStatus, 'New');
+        $membershipParams['status_id'] = $calcStatus['id'] ?? 'New';
```
Comment (colemanw) [inline CRM/Contribute/BAO/Contribution.php:4259]: Yay. One less of those nasty things.

**Pull request merged**